crates/block-producer/src/withdrawal.rs

```diff
@@ -10,148 +9,209 @@ use gw_types::{
+    ) -> Self {
+        let mut total_sudt_capacity = 0u64;
+        let mut sudt_custodians = HashMap::new();
+
+        for (sudt_type_hash, (balance, type_script)) in available_custodians.sudt.into_iter() {
+            let (change, _data) =
+                generate_finalized_custodian(rollup_context, balance, type_script.clone());
+            let change_capacity: u64 = change.capacity().unpack();
+            total_sudt_capacity = total_sudt_capacity.saturating_add(change_capacity);
+            sudt_custodians.insert(sudt_type_hash, (change_capacity, balance, type_script));
         }
-    );
-
-    let custodian_cells = rpc_client
-        .query_finalized_custodian_cells(&total_withdrawals_amount, last_finalized_block_number)
-        .await?;
-    assert!(custodian_cells.fullfilled_sudt_script.len() == total_withdrawals_amount.sudt.len());
-
-    let build_withdraw_output = |req: WithdrawalRequest| -> Result<(CellOutput, Bytes)> {
-        let withdrawal_capacity: u64 = req.raw().capacity().unpack();
-        let lock_args: Bytes = {
-            let withdrawal_lock_args = WithdrawalLockArgs::new_builder()
-                .account_script_hash(req.raw().account_script_hash())
-                .withdrawal_block_hash(block.hash().pack())
-                .withdrawal_block_number(block.raw().number())
-                .sudt_script_hash(req.raw().sudt_script_hash())
-                .sell_amount(req.raw().sell_amount())
-                .sell_capacity(withdrawal_capacity.pack())
-                .owner_lock_hash(req.raw().owner_lock_hash())
-                .payment_lock_hash(req.raw().payment_lock_hash())
-                .build();
 
-            let rollup_type_hash = rollup_context.rollup_script_hash.as_slice().iter();
-            rollup_type_hash
-                .chain(withdrawal_lock_args.as_slice().iter())
-                .cloned()
-                .collect()
+        let ckb_custodian_min_capacity = {
+            let lock = build_finalized_custodian_lock(rollup_context);
+            (8 + lock.as_slice().len() as u64) * 100000000
         };
 
-        let lock = Script::new_builder()
-            .code_hash(rollup_context.rollup_config.withdrawal_script_type_hash())
-            .hash_type(ScriptHashType::Type.into())
-            .args(lock_args.pack())
-            .build();
+        let ckb_custodian_capacity = available_custodians
+            .capacity
+            .saturating_sub(total_sudt_capacity);
+        let ckb_balance = ckb_custodian_capacity.saturating_sub(ckb_custodian_min_capacity);
+        let ckb_custodian = (
+            ckb_custodian_capacity,
+            ckb_balance,
+            ckb_custodian_min_capacity,
+        );
+
+        Generator {
+            rollup_context,
+            ckb_custodian,
+            sudt_custodians,
+            withdrawals: Default::default(),
+        }
+    }
 
+    pub fn insert(&mut self, req: &WithdrawalRequest, block: &L2Block) -> Result<()> {
```

Comment:
Rename to `push_and_verify` or something similar, the name `insert` does not reflect the purpose of this function.

Reply:
rename to ```include_and_verify```.
Done